lib/openhab/core/items/item.rb

```diff
@@ -28,6 +28,25 @@ def ===(other)
         #   The item's descriptive label.
         #   @return [String, nil]
 
+        #
+        # @!attribute [rw] tags
+        #   The item's tags
+        #   @return [Array<String>]
+        #   @overload tags
+        #     Returns the item's tags.
+        #     @return [Array<String>]
+        #   @overload tags=(*tags)
+        #     Sets the item's tags
+        #     @param [String,Symbol,Semantics::Tag] tags Tags to set.
+        #     @return [void]
+
+        # @!visibility private
+        def tags=(*tags)
```

Comment:
plz to raise an error (like in `Items::Metadata::Hash#provider`) unless `Provider.registry.provider_for(self).is_a?(ManagedProvider)`. the `update()` call below won't work on a non-managed provider, and will just super-confuse the state of things if you try to add tags on an item from GenericItemProvider

you should also do the check between preferredProvider and provider, warning against modifying tags on items not owned by this script

Reply:
I will add this guard, however, I'm just wondering:

Why can't we:
- Remove items provided by GenericItemProvider (non managed)?
- Add / alter tags of such items

IMO, you should be able to do the above.  It's technically not wrong to do so. It simply removes the item from the itemregistry after `remove` is called. Yes, the removal / tags modifications aren't persistent, and when openhab reloads the items file, the item will come back to life and tags will reset, but should we stop people from doing it? At most, we could issue a warn in the log, with a way to silence it.

I can't think of a scenario for tags - this is probably far-fetched, but perhaps someone wants to modify an item's semantics dynamically based on various runtime conditions?

As for the `update()` call, it could be something like:
`Provider.registry.provider_for(self)&.update(self)` i.e. don't call update if the provider is nil.

Reply:
* It is _not_ possible to remove items provided by GenericItemProvider. It does not have a remove method. You can call remove() on ItemRegistry passing in an Item provided by the GenericItemProvider, but Registries simply forward any add/update/remove requests to the default ManagedProvider. Which will simply ignore the request, because it does not recognize the Item.
 * While the Java classes will allow you to modify the tags of _any_ item directly, it's not intended to be used for mucking about with items your code doesn't own. For starters, if the provider that owns it doesn't send an updated event through the registry, other components won't pro-actively notice any changes (like if you have a rule on an item changing, or the HomeKit addon used to use tags for configuration, and it wouldn't know to reload). It also means that you would be left in an inconsistent state, where the in-memory representation of the item does not match what's on disk. You'll notice MainUI does _not_ allow you to modify tags for file-based items (or items from JRuby scripts either...).

I'm very adamant to prevent issues here. One example of issues that can occur is if an Item is provided by a managed provider, and then _also_ provided by another (generic) provider, the registry will silently refuse the addition, but the provider thinks it worked. When the generic provider then notifies that it removed the item, the registry doesn't realize it never recorded this one, and will instead remove the managed provider's instance of the item. At which point, your item is gone, but you can't _re-add_ it via the managed provider, because the managed provider thinks it's there and will refuse to add it. The only way to recover is to reboot openHAB. There are various related issues here too - the inability to remove an item also happens.

Reply:
I feel that we are sugarcoating things too much. I feel that the provider check *and* the call to `update` are unnecessary and if they do need to be done, they need to be done in core, within the implementation of `addTags` and `removeAllTags` themselves.